naptime-testing/src/test/scala/org/coursera/naptime/NestedMacroCourierTests.scala

```diff
@@ -0,0 +1,95 @@
+package org.coursera.naptime
+
+import com.linkedin.data.schema.DataSchema
+import com.linkedin.data.schema.RecordDataSchema
+import org.coursera.naptime.model.Keyed
+import org.coursera.naptime.resources.TopLevelCourierCollectionResource
+import org.coursera.naptime.router2.Router
+import org.junit.Test
+import org.scalatest.junit.AssertionsForJUnit
+
+/**
+ * This test suite uses Courier to exercise advanced use cases for Naptime.
+ */
+object NestedMacroCourierTests {
+
+  class CoursesResource extends TopLevelCourierCollectionResource[String, Course] {
+    override def resourceName: String = "courses"
+
+    override implicit lazy val Fields: Fields[Course] = BaseFields.withRelated("instructors" -> ResourceName("instructors", 1))
+
+    var relatedFun: Option[String => List[String]] = None
+
+    def fallbackRelatedFunction(id: String): List[String] = {
+      id match {
+        case "abc" => List("123")
+        case "xyz" => List.empty
+        case "qrs" => List("456", "789")
+        case _ => List.empty
+      }
+    }
+
+    private[this] def makeCourse(id: String): Keyed[String, Course] = {
+      Keyed(id,
+        Course(s"course-$id", s"$id description", relatedFun.map(_(id)).getOrElse(fallbackRelatedFunction(id))))
+    }
+
+    def get(id: String) = Nap.get { ctx =>
+      Ok(makeCourse(id))
+    }
+
+    def multiGet(ids: Set[String]) = Nap.multiGet { ctx =>
+      Ok(ids.map(id => makeCourse(id)).toList)
+    }
+
+    def search(query: Option[String]) = Nap.finder { ctx =>
+      val ids = query.map(List(_)).getOrElse(List("abc", "qrs"))
+      Ok(ids.map(id => makeCourse(id)))
+    }
+
+    def byInstructor(instructorId: String) = Nap.finder { ctx =>
+      val ids = instructorId match {
+        case "123" => List("xyz")
+        case "456" => List("xyz", "abc")
+        case "789" => List("qrs")
+        case _ => List.empty
+      }
+      Ok(ids.map(id => makeCourse(id)))
+    }
+  }
+
+  class InstructorsResource extends TopLevelCourierCollectionResource[String, Instructor] {
+    override def resourceName: String = "instructors"
+
+    def multiGet(ids: Set[String]) = Nap.multiGet { ctx =>
+      ???
+    }
+
+  }
+
+  val courseRouter = Router.build[CoursesResource]
+  val instructorRouter = Router.build[InstructorsResource]
+}
+
+class NestedMacroCourierTests extends AssertionsForJUnit {
+
+  @Test
+  def checkCoursesMergedType(): Unit = {
```

Comment:
this test looks good. I wonder if there's any way though to write an expected courier model, and assert that the generated one and the hard-coded one are equal?

that way you could just write something like:
```
record MergedCourse {

  id: String

  @related("instructors.v1")
  instructors: [InstructorId]

  ...

}
```
and somehow assert that that is equal to the generated one.  it might make the tests more maintainable and easier to comprehend what's going on